internal/rest/resources/resources.go

```diff
@@ -44,68 +45,97 @@ var InternalEndpoints = rest.Resources{
 		}
-	}
 
-	return nil
-}
+		if server.ServeUnix && !server.CoreAPI {
+			return fmt.Errorf("Cannot serve non-core API resources over the core unix socket")
+		}
 
-// GetAndValidateCoreEndpoints extracts all endpoints from extensionServers that should be allocated to the core listener.
-// It also performs the following validations:
-// 1. Only one core API server is allowed.
-// 2. Server configurations are properly set.
-// 3. Path prefixes for endpoints belonging to a single server are not duplicated
-// 4. Enpoints defined in extensionServers do not conflict with internal endpoints.
-func GetAndValidateCoreEndpoints(extensionServers []rest.Server) ([]rest.Resources, error) {
-	var coreEndpoints []rest.Resources
-	var numCoreAPIs int
-
-	for _, extensionServer := range extensionServers {
-		if !extensionServer.CoreAPI {
-			continue
+		if server.CoreAPI && server.Certificate != nil {
+			return fmt.Errorf("Core API server cannot have a pre-defined certificate")
 		}
 
-		numCoreAPIs++
-		if numCoreAPIs > 1 {
-			return nil, fmt.Errorf("Only one core API server is allowed")
+		if server.CoreAPI && server.Address != (types.AddrPort{}) {
+			return fmt.Errorf("Core API server cannot have a pre-defined address")
+		}
+
+		// Ensure all servers with a defined address are unique.
+		if server.Address != (types.AddrPort{}) {
+			if serverAddresses[server.Address.String()] {
+				return fmt.Errorf("Server address %q conflicts with another Server", server.Address.String())
+			}
+
+			serverAddresses[server.Address.String()] = true
+		} else if server.Protocol != "" {
+			return fmt.Errorf("Server protocol defined without address")
 		}
 
-		err := extensionServer.ValidateServerConfigs()
+		// Ensure no endpoint path conflicts with another endpoint on the same server.
+		// If a server lacks an address, we need to compare it to every other server
+		// that also lacks an address, as well as the internal endpoints.
+		serverPaths, err := resourcesConflict(server, existingEndpointPaths)
 		if err != nil {
-			return nil, err
+			return err
 		}
 
-		seen := make(map[string]bool)
-		for _, endpoints := range extensionServer.Resources {
-			if seen[string(endpoints.PathPrefix)] {
-				return nil, fmt.Errorf("Path prefix %q is duplicated in server configuration", endpoints.PathPrefix)
+		// Update the existing paths now that we have checked a new Server.
+		for k, v := range serverPaths {
+			if !existingEndpointPaths[k] {
+				existingEndpointPaths[k] = v
 			}
+		}
+	}
 
-			err = checkInternalEndpointsConflict(endpoints)
-			if err != nil {
-				return nil, err
-			}
+	return nil
+}
 
-			coreEndpoints = append(coreEndpoints, endpoints)
-			seen[string(endpoints.PathPrefix)] = true
+// resourcesConflict returns an error if the endpoint paths of the given server conflict with any paths in the given map of existing paths.
+func resourcesConflict(server rest.Server, existingPaths map[string]bool) (map[string]bool, error) {
```

Comment:
@masnax just a thought here, if we restrict the number of servers with `CoreAPI=true` to 1, then we wouldn't need to do cross server validation here? From a usage perspective I can't really see a reason why multiple server structs would be defined with `CoreAPI=true`. Although I think maybe we also shouldn't assume what the user might want to do, so feel free to disregard this comment :)

Reply:
I'm thinking with `PreInit` and `ServeUnix` there may be some cases of multiple core API servers with different combinations of those 2 flags.